Config

R5dan opened the pull request:
Supersedes #2209 with better implementation.

Changes:
- Deprecates all `proxy`, `session` and `timeout` in functions
- Adds `yf.set_config`
- Allows setting:
    - `proxy`
    - `session`
    - `timout`
    - `url`
    - `language`
    - `region`
- Adds deprecator for future use 
- Adds docs for use of `yf.set_config`

R5dan commented:
@ValueRaider what do you think? Happy to move `yf.set_config` if you can think of somewhere. I have deprecated all of the `session`, `proxy` and `timeout` (thats why its adding code not removing it but that can all go away in a few versions time)

R5dan commented:
@ValueRaider What do you think?

ValueRaider commented:
I haven't had time to try and review yet.

ValueRaider commented:
Remove `removed_in`. I generally remove a few releases after warning was added.

Is `**message_params` actually used anywhere? If not, remove it.

Set `new` on everything, so user is told the new function or argument.

R5dan commented:
> * remove `since` as well, how can developer always know which version their PR will end up in?

Unless it is near the next release cycle, it is normally pretty obvious, plus it can always be changed - like I'm going to have to do now. I think this should stay as (1) more information never hurt - even if just for devs and (2) it lets the user know when it was deprecated, so if they really wanted to or needed to could revert back to that until they change it

> * removed_in

I think this should probably stay as it tells them exactly when to change by.
With this PR I would also like to add a 2 or 3 release gap between all deprecations and removals as I believe `earnings` in `fundamentals` has been deprecated since `0.2.41`



> * is `**message_params` actually used anywhere? If not, remove it.

I don't believe it is however think it should stay for future use.

> User should not be informed on internal deprecations that aren't their fault

Yes, I agree. I marked the "getters" with deprecate as they weren't the common "private attributes"

> > dat.balance_sheet
> yfinance/scrapers/fundamentals.py:99: DeprecationWarning: Parameter 'proxy' of Financials.get_financials_time_series is deprecated

That shouldn't raise a `Deprecation Warning` as it doesn't set `proxy`. Will fix.

> Set `new` on everything, so user is told the new function or argument.

Not possible for `proxy` `session` or `timeout` as they are simply removed. I don't believe I am deprecating anything that have replacements.

> What happens if user changes session with `set_config` after `YfData` singleton already created?

In my initial testing of it, it seemed to work. However it doesn't seem to set it if its called after. Will have to fix. Any ideas how to?

I will change the default of timeout

ValueRaider commented:
OK keep `since`. But `removed_in` is still a problem for us developers - the gap is determined by time more than number of releases and [we don't have a regular release schedule](https://github.com/ranaroussi/yfinance/tags).

> `**message_params` should stay for future use

Without a current use case it isn't immediately obvious what it does. I figured it out and saw a bug, it's broken. Also the variable name could be more specific.

> Not possible for proxy session or timeout as they are simply removed. I don't believe I am deprecating anything that have replacements.

You need to inform users of the replacement in that deprecation message. That's my point.

Move `set_config` to the `YfData` class like a normal setter function.

Run the unit tests to check for bugs. You might need to rebase to latest `dev` (not merge) to pull in fixes.

ValueRaider commented:
Immediately after the rebase (done on your local system), do a force push to your fork #1084

R5dan commented:
Now I get:
```
Pushing to https://github.com/R5dan/yfinance-package.git
To https://github.com/R5dan/yfinance-package.git
 ! [rejected]        ConfigV2 -> ConfigV2 (stale info)
```
(Sorry Im not very good with git)

R5dan commented:
@ValueRaider How is this?

ValueRaider commented:
How are you testing this?

R5dan commented:
@ValueRaider sorry for the late reply. I updated the tests and have been running them

ValueRaider commented:
```
import yfinance as yf
yf.set_config()
```

R5dan commented:
> Do you use this branch as your daily driver?

No

> Bug:
> 
> ```
> import yfinance as yf
> yf.set_config()
> ```
> 
> > NameError: name 'requests' is not defined

Fixed

> Then with `yf.Ticker('NVDA').balance_sheet`, user gets a wall of warning messages that aren't directly related to their code:

They were debug prints I forgot to remove. Removed now.

R5dan commented:
> Would highly suggest you to use and test your own branch.

I do test them, just not thoroughly enough obviously



> ### Bug:
> ```python
> import yfinance as yf
> 
> yf.Ticker("AAPL").history()
> ```
> 
> ```python
> C:\Users\dhruv\yfinance\yfinance\base.py:86: DeprecationWarning: `proxy` is deprecated. Please set it using `yf.set_config`
> Deprecated since version 0.2.53
> `timeout` is deprecated. Please set it using `yf.set_config`
> Deprecated since version 0.2.53
>   self._price_history = PriceHistory(self._data, self.ticker, self._get_ticker_tz(self.proxy, timeout=10))
> $AAPL: possibly delisted; no price data found  (period=1mo)
> 
> Empty DataFrame
> Columns: [Open, High, Low, Close, Adj Close, Volume]
> Index: []
> ```

Weird, it seems like `history` is getting called twice - or more the `deprecated` wrapper is getting called twice but I can't find from where. One of them is being called with: `{'self': yfinance.Ticker object <AAPL>, 'proxy': None, 'timeout': 10}` which causes `timeout` and `proxy` to be flagged and another with what it should be `{'self': <yfinance.scrapers.history.PriceHistory object at 0x00000254A69AACF0>}`

dhruvan2006 commented:
> Weird, it seems like `history` is getting called twice

I think you need to search the whole codebase and remove every occurrence of passing proxy, session, etc. down to `YfData`

R5dan commented:
> I think you need to search the whole codebase and remove every occurrence of passing proxy, session, etc. down to `YfData`

I did initially, but @ValueRaider told me to deprecate them instead.

ValueRaider commented:
> or more the deprecated wrapper is getting called twice but I can't find from where

Would *removing* the deprecated wrapper from `scrapers/*` help? Only need them on the public-facing functions. If users are directly calling internal functions, they can cope with the arguments suddenly disappearing (in future release).

> I think you need to search the whole codebase and remove every occurrence of passing proxy, session, etc. down to YfData

My philosophy is similar to Pandas - warn users of breaking changes early, gives them time to adjust their codes. No one wants their algo-trading to suddenly break.

ValueRaider commented:
> functions like history are just run the function with args and kwargs

This was just me being lazy back when I moved the fetch logic into `scrapers/history.py`. It also breaks the auto-docs. Can you copy over the actual arguments?

R5dan commented:
> This was just me being lazy back when I moved the fetch logic into `scrapers/history.py`. It also breaks the auto-docs. Can you copy over the actual arguments?

I thought I had mentioned this before and you said no. Anyway added now.

Also I fixed the bug

ValueRaider commented:
Better because fewer messages, but the user is still being shown internal issues:

```
>>> yf.Ticker('NVDA').balance_sheet
yfinance/scrapers/fundamentals.py:99: DeprecationWarning: 
`proxy` is deprecated. Please set it using `yf.set_config`
Deprecated since version 0.2.53
  return self.get_financials_time_series(timescale, keys, proxy)
```

R5dan commented:
I thought I had removed them all. Hows this @ValueRaider? Also how do you get these? I don't see these when I run tests

ValueRaider commented on doc/source/advanced/config.rst:
```diff
@@ -0,0 +1,20 @@
+    yf.set_config(proxy=None, timeout=30, lang="en-US", region="US", session=None, url="finance.yahoo.com")
+
+.. important::
+    DO NOT ADD `https://` to start of url. This will cause an error.
```
Can we search-delete for `https://`?

R5dan replied:
Yes, good idea. I will impliment

ValueRaider commented on yfinance/scrapers/quote.py:
```diff
@@ -584,30 +581,18 @@ def _fetch(self, proxy, modules: list):
         except requests.exceptions.HTTPError as e:
             utils.get_yf_logger().error(str(e))
             return None
         return result
```
why you delete this?

R5dan replied:
I moved all urls to `YfData.URLS` to work with the custom urls - (having a different base)

ValueRaider commented on yfinance/data.py:
```diff
@@ -398,21 +521,19 @@ def _make_request(self, url, request_method, user_agent_headers=None, body=None,
 
         return response
 
-    @lru_cache_freezeargs
-    @lru_cache(maxsize=cache_maxsize)
-    def cache_get(self, url, user_agent_headers=None, params=None, proxy=None, timeout=30):
-        return self.get(url, user_agent_headers, params, proxy, timeout)
```
why you delete this?

R5dan replied:
The `lru_cache_freezeargs` made it so I couldn't update the params with the region and lang but I am now testing a workaround

ValueRaider commented:
> I don't think I have

Why? Looks good to me.

R5dan commented:
Because of the recommiting of everything

ValueRaider commented:
That's github - `git log` shows original dates.

R5dan commented:
@ValueRaider how is this?

ValueRaider commented:
Broken. Run the tests.

R5dan commented:
I get no errors I don't get on the `dev` branch.

I get the following errors on both branches:
- Unable to open database
- Invalid url, usually caused by an invalid crumb I think

I do not get these errors when running code in yfinance only the tests

Could you specify what errors you get?

ValueRaider commented:
> yf.Ticker('NVDA').income_stmt

Returns empty table, obv should return a full table.

R5dan commented:
Weird, I get a full table:

```
                                                        2025-01-31     2024-01-31     2023-01-31     2022-01-31
Tax Effect Of Unusual Items                                    0.0            0.0   -284130000.0            0.0
Tax Rate For Calcs                                           0.133           0.12           0.21          0.019
Normalized EBITDA                                    86137000000.0  35583000000.0   7340000000.0  11351000000.0
Total Unusual Items                                            0.0            0.0  -1353000000.0            0.0
Total Unusual Items Excluding Goodwill                         0.0            0.0  -1353000000.0            0.0
Net Income From Continuing Operation Net Minori...   72880000000.0  29760000000.0   4368000000.0   9752000000.0
Reconciled Depreciation                               1864000000.0   1508000000.0   1544000000.0   1174000000.0
Reconciled Cost Of Revenue                           32639000000.0  16621000000.0  11618000000.0   9439000000.0
EBITDA                                               86137000000.0  35583000000.0   5987000000.0  11351000000.0
EBIT                                                 84273000000.0  34075000000.0   4443000000.0  10177000000.0
Net Interest Income                                   1539000000.0    609000000.0      5000000.0   -207000000.0
Interest Expense                                       247000000.0    257000000.0    262000000.0    236000000.0
Interest Income                                       1786000000.0    866000000.0    267000000.0     29000000.0
Normalized Income                                    72880000000.0  29760000000.0   5436870000.0   9752000000.0
Net Income From Continuing And Discontinued Ope...   72880000000.0  29760000000.0   4368000000.0   9752000000.0
Total Expenses                                       49044000000.0  27950000000.0  21397000000.0  16873000000.0
Total Operating Income As Reported                   81453000000.0  32972000000.0   4224000000.0  10041000000.0
Diluted Average Shares                               24804000000.0  24940000000.0  25070000000.0  25350000000.0
Basic Average Shares                                 24555000000.0  24690000000.0  24870000000.0  24960000000.0
Diluted EPS                                                   2.94           1.19           0.17          0.385
Basic EPS                                                     2.97           1.21           0.18          0.391
Diluted NI Availto Com Stockholders                  72880000000.0  29760000000.0   4368000000.0   9752000000.0
Net Income Common Stockholders                       72880000000.0  29760000000.0   4368000000.0   9752000000.0
Net Income                                           72880000000.0  29760000000.0   4368000000.0   9752000000.0
Net Income Including Noncontrolling Interests        72880000000.0  29760000000.0   4368000000.0   9752000000.0
Net Income Continuous Operations                     72880000000.0  29760000000.0   4368000000.0   9752000000.0
Tax Provision                                        11146000000.0   4058000000.0   -187000000.0    189000000.0
Pretax Income                                        84026000000.0  33818000000.0   4181000000.0   9941000000.0
Other Income Expense                                  1034000000.0    237000000.0  -1401000000.0    107000000.0
Other Non Operating Income Expenses                   1034000000.0    237000000.0    -48000000.0    107000000.0
Special Income Charges                                         0.0            0.0  -1353000000.0            0.0
Restructuring And Mergern Acquisition                          0.0            0.0   1353000000.0            0.0
Net Non Operating Interest Income Expense             1539000000.0    609000000.0      5000000.0   -207000000.0
Interest Expense Non Operating                         247000000.0    257000000.0    262000000.0    236000000.0
Interest Income Non Operating                         1786000000.0    866000000.0    267000000.0     29000000.0
Operating Income                                     81453000000.0  32972000000.0   5577000000.0  10041000000.0
Operating Expense                                    16405000000.0  11329000000.0   9779000000.0   7434000000.0
Research And Development                             12914000000.0   8675000000.0   7339000000.0   5268000000.0
Selling General And Administration                    3491000000.0   2654000000.0   2440000000.0   2166000000.0
Gross Profit                                         97858000000.0  44301000000.0  15356000000.0  17475000000.0
Cost Of Revenue                                      32639000000.0  16621000000.0  11618000000.0   9439000000.0
Total Revenue                                       130497000000.0  60922000000.0  26974000000.0  26914000000.0
Operating Revenue                                   130497000000.0  60922000000.0  26974000000.0  26914000000.0
```

R5dan commented:
Fixed

R5dan commented:
Anything else needed @ValueRaider ?

ValueRaider commented:
Every time I find a bug after 5 seconds, just using the API. https://yfinance-python.org/reference/index.html

dhruvan2006 commented:
Is this PR being abandoned?

Maybe I can suggest taking this on in a new PR?

R5dan commented:
No not abandoned, just testing. I am currently getting a different result in the Market, sorry it is taking so long: will focus on it more.

R5dan commented:
Currently there is a problem with the crumbs, I think its because of the cookies. But I havent changed any of the logic

R5dan commented:
If either of you two find any bugs please let me know. The bug with the invalid crumb is weird because it only occurs during tests: but when I use the exact same code in production it works

R5dan commented:
@ValueRaider I have tested this with a few packages and find no bugs.

dhruvan2006 commented:
This PR is not backwards compatible. It doesn't respect timeouts passed. This code runs without throwing an Exception.

```python
import yfinance as yf

yf.download("AAPL", timeout=1e-8)
```

R5dan commented:
From what I got, entirely intentional functionality: there should be a `Deprecation` error

dhruvan2006 commented:
Run the same code on your branch and on `dev`/`main`

R5dan commented:
I see what you mean, had an or statement the wrong way. Fixed now

ValueRaider commented:
Still errors:
```
>>> dat.earnings_dates
  File "yfinance/yfinance/data.py", line 560, in _make_request
    raise ValueError("Don't manually add 'region' to params dict, let data.py handle it")
```

This is silly now, someone else can implement.

R5dan commented:
I agree, this pr is a mess: some of my worse code. But it just needs bug fixes, which I think will be easier than redoing it. Doesn't help that the tests don't work, so I can't run them: is there a better way to do it?

ValueRaider commented:
Too much engineering imo, also you forgot that the internal functions can completely discard proxy etc.

Compare to PR #2391 - the core change is simply https://github.com/ranaroussi/yfinance/pull/2391/files#diff-98649f0c0b57881da326077ecbb9f45ec98808e31a6e6ad9206c0e42d2f5b2c9
everything else is just deprecations and interceptions

R5dan commented:
Ok, I kind of see your point.

There was a few complex features, such as the urls etc.

I still think that all the feature in this PR are useful, just might need to be less complex

ValueRaider commented:
Try splitting and merge into dev incrementally. Adopt philosophy that every LoC you change is important.
